### PR TITLE
Added a few methods to the public API of collections

### LINKED ISSUE #1377: TreeMap.narrow() is not public
Are there other static methods in impl classes that are package private but should be public?

### PR DESCRIPTION
**Pull request by paplorinc**
Fixes #1377

### CODE REVIEW

File: javaslang/src/main/java/javaslang/Lazy.java [hunk 1]
```diff
@@ -68,7 +61,7 @@ private Lazy(Supplier<? extends T> supplier) {
      * @return the given {@code lazy} instance as narrowed type {@code Lazy<T>}.
      */
     @SuppressWarnings("unchecked")
-    static <T> Lazy<T> narrow(Lazy<? extends T> lazy) {
+    public static <T> Lazy<T> narrow(Lazy<? extends T> lazy) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/Lazy.java]: There's no test for this

File: javaslang/src/main/java/javaslang/collection/Array.java [hunk 2]
```diff
@@ -127,7 +127,7 @@ private Array(Object[] back) {
      * @param <T>        Component type of the Stream.
      * @return A Array containing the given elements in the same order.
      */
-    static <T> Array<T> ofAll(java.util.stream.Stream<? extends T> javaStream) {
+    public static <T> Array<T> ofAll(java.util.stream.Stream<? extends T> javaStream) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Array.java]: @ruslansennov

File: javaslang/src/main/java/javaslang/collection/Array.java [hunk 3]
```diff
@@ -482,7 +482,7 @@ private Array(Object[] back) {
      * @return an Array with the values built up by the iteration
      * @throws IllegalArgumentException if {@code f} is null
      */
-    static <T,U> Array<U> unfoldRight(T seed, Function<? super T,Option<Tuple2<? extends U, ? extends T>>> f) {
+    public static <T, U> Array<U> unfoldRight(T seed, Function<? super T, Option<Tuple2<? extends U, ? extends T>>> f) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Array.java]: @emmanueltouzery

File: javaslang/src/main/java/javaslang/collection/CharSeq.java [hunk 4]
```diff
@@ -1841,14 +1840,14 @@ public int lastIndexOf(int ch, int fromIndex) {
      * @param fromIndex the index to start the search from.
      * @return {@code Some(index)} or {@code None} if not found.
      */
-    Option<Integer> lastIndexOfOption(int ch, int fromIndex) {
+    public Option<Integer> lastIndexOfOption(int ch, int fromIndex) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/CharSeq.java]: @paplorinc

File: javaslang/src/main/java/javaslang/collection/CharSeq.java [hunk 5]
```diff
@@ -2500,12 +2501,12 @@ public CharSeq trim() {
     }
 
     @FunctionalInterface
-    interface CharUnaryOperator {
+    public interface CharUnaryOperator {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/CharSeq.java]: @danieldietrich

File: javaslang/src/main/java/javaslang/collection/TreeMap.java [hunk 6]
```diff
@@ -108,7 +108,7 @@ private TreeMap(RedBlackTree<Tuple2<K, V>> entries) {
      * @return the given {@code treeMap} instance as narrowed type {@code TreeMap<K, V>}.
      */
     @SuppressWarnings("unchecked")
-    static <K, V> TreeMap<K, V> narrow(TreeMap<? extends K, ? extends V> treeMap) {
+    public static <K, V> TreeMap<K, V> narrow(TreeMap<? extends K, ? extends V> treeMap) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/TreeMap.java]: https://github.com/javaslang/javaslang/issues/1377

**Comment by emmanueltouzery** [comment]
regarding the unfold* they were meant to be public. I missed it because I first added them to `List` and that's an interface, no need for the `public` there. I should have added the `public` when implementing it in classes, but I missed it. So yes, it should be made public.

**Comment by emmanueltouzery** [comment]
hmm wait this is static. Maybe the public would be needed also for interfaces like List then?

**Comment by paplorinc** [comment]
> So yes, it should be made public

Thanks :)

> Maybe the public would be needed also for interfaces like List then

Interface methods can only be public, no need to specify it explicitly.

**Comment by danieldietrich** [comment]
Very nice! The branch has conflict now. Could you please solve them an I will merge. Thanks!

**Comment by paplorinc** [comment]
Done, please recheck!

**Comment by danieldietrich** [comment]
Great, thank you!

**Pull request merged**